accounts/usbwallet/wallet.go

```diff
@@ -524,6 +526,45 @@ func (w *wallet) signHash(account accounts.Account, hash []byte) ([]byte, error)
 
 // SignData signs keccak256(data). The mimetype parameter describes the type of data being signed
 func (w *wallet) SignData(account accounts.Account, mimeType string, data []byte) ([]byte, error) {
+
+	// dispatch to 712 signing if the mimetype is TypedData and the format matches
+	if mimeType == accounts.MimetypeTypedData && data[0] == 0x19 && data[1] == 0x01 && len(data) == 66 {
+		w.stateLock.RLock() // Comms have own mutex, this is for the state fields
+		defer w.stateLock.RUnlock()
+
+		// If the wallet is closed, abort
+		if w.device == nil {
+			return nil, accounts.ErrWalletClosed
+		}
+		// Make sure the requested account is contained within
+		fmt.Printf("looking for %x\n", account.Address)
+		path, ok := w.paths[account.Address]
+		if !ok {
+			return nil, accounts.ErrUnknownAccount
+		}
+		// All infos gathered and metadata checks out, request signing
+		<-w.commsLock
+		defer func() { w.commsLock <- struct{}{} }()
+
+		// Ensure the device isn't screwed with while user confirmation is pending
+		// TODO(karalabe): remove if hotplug lands on Windows
+		w.hub.commsLock.Lock()
+		w.hub.commsPend++
+		w.hub.commsLock.Unlock()
+
+		defer func() {
+			w.hub.commsLock.Lock()
+			w.hub.commsPend--
+			w.hub.commsLock.Unlock()
+		}()
+		// Sign the transaction
+		signature, err := w.driver.SignTypedMessage(path, data[2:34], data[34:66])
+		if err != nil {
+			return nil, err
+		}
+		return signature, nil
+	}
```

Comment:
This whole clause is more or less a copy-paste of `SignTx`. Is there any way it can be refactored so they both reuse the whole dance around `commsLock` and account-lookup and stuff, and not have the same code duplicated?

Reply:
Hmm, I'm not sure exactly how to do this, since the `defer`ed functions are relative to the scope of the surrounding function. Do you have an idea?